File: frontend/src/types/creator.ts
export type Platform = 'douyin' | 'xiaohongshu'
export type Priority = 'high' | 'normal' | 'low'
export type MonitoringStatus = 'active' | 'paused'
export type CollectorType = 'mock' | 'douyin_public_web' | 'tikhub_douyin'
export type MonitorScope = 'creator_collection' | 'single_content'
export type DataQualityStatus = 'pending' | 'mock' | 'verified' | 'partial' | 'failed'
export type ContentCollectionStatus =
  | 'pending'
  | 'success'
  | 'partial'
  | 'unavailable'
  | 'baseline_created'
  | 'no_new_content'
  | 'metrics_refreshed'
  | 'budget_limited'
  | 'failed'

export interface Creator {
  id: number
  platform: Platform
  platform_account_id: string
  platform_display_id: string | null
  nickname: string
  profile_url: string
  avatar_url: string | null
  bio: string | null
  verified_info: string | null
  location: string | null
  group_name: string | null
  tags: string[]
  priority: Priority
  monitor_interval_minutes: number
  monitor_scope: MonitorScope
  monitoring_status: MonitoringStatus
  collector_type: CollectorType
  collector_version: string | null
  data_quality_status: DataQualityStatus
  last_content_status: ContentCollectionStatus
  last_collection_error: string | null
  content_baseline_established_at: string | null
  follower_count: number
  following_count: number
  total_like_count: number
  content_count: number
  last_collected_at: string | null
  next_collect_at: string | null
  consecutive_failures: number
  created_at: string
  updated_at: string
}

export interface CreatorPayload {
  platform: Platform
  platform_account_id: string
  platform_display_id?: string | null
  nickname: string
  profile_url: string
  avatar_url?: string | null
  bio?: string | null
  verified_info?: string | null
  location?: string | null
  group_name?: string | null
  tags: string[]
  priority: Priority
  monitor_interval_minutes: number
  monitor_scope?: MonitorScope
  collector_type: CollectorType
  follower_count?: number
  following_count?: number
  total_like_count?: number
  content_count?: number
  profile_resolved?: boolean
}

export interface CreatorProfileResolvePayload {
  platform: Platform
  input_value: string
}

export interface CreatorProfileResolveResult {
  platform: Platform
  platform_account_id: string
  platform_display_id: string | null
  nickname: string
  profile_url: string
  avatar_url: string | null
  bio: string | null
  verified_info: string | null
  location: string | null
  follower_count: number
  following_count: number
  total_like_count: number
  content_count: number
  collector_type: CollectorType
  sec_user_id: string | null
  warnings: string[]
}

export interface CreatorListResponse {
  items: Creator[]
  total: number
  page: number
  page_size: number
}

export interface CreatorSnapshot {
  id: number
  creator_id: number
  follower_count: number
  following_count: number
  total_like_count: number
  content_count: number
  collector_type: CollectorType
  data_quality_status: DataQualityStatus
  captured_at: string
}

export interface CollectionRun {
  id: number
  creator_id: number
  task_type: string
  status: string
  trigger_source: 'initial' | 'manual' | 'scheduled'
  attempt: number
  collector_type: CollectorType | null
  error_type: string | null
  duration_ms: number | null
  started_at: string
  finished_at: string | null
  error_message: string | null
  result_summary: Record<string, unknown> | null
  creator: {
    id: number
    nickname: string
    platform: Platform
  }
}

export interface CollectionResult {
  creator: Creator
  snapshot: CreatorSnapshot
  run: CollectionRun
}

export interface CollectionRetryQueued {
  creator_id: number
  task_id: string
  status: 'queued'
  retry_after_seconds: number
  message: string
}

export interface CollectionRunListResponse {
  items: CollectionRun[]
  total: number
  page: number
  page_size: number
}

File: backend/app/schemas/creator.py
from datetime import datetime
from typing import Literal, Self

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

from app.schemas.base import UtcResponseModel
from app.utils.profile_urls import normalize_profile_url

Platform = Literal["douyin", "xiaohongshu"]
Priority = Literal["high", "normal", "low"]
MonitoringStatus = Literal["active", "paused"]
CollectorType = Literal["mock", "douyin_public_web", "tikomni_douyin"]


class CreatorCreate(BaseModel):
    platform: Platform
    platform_account_id: str = Field(min_length=1, max_length=128)
    platform_display_id: str | None = Field(default=None, max_length=128)
    nickname: str = Field(min_length=1, max_length=128)
    profile_url: str = Field(min_length=1, max_length=500)
    avatar_url: str | None = Field(default=None, max_length=1000)
    bio: str | None = None
    verified_info: str | None = Field(default=None, max_length=255)
    location: str | None = Field(default=None, max_length=128)
    group_name: str | None = Field(default=None, max_length=128)
    tags: list[str] = Field(default_factory=list)
    priority: Priority = "normal"
    monitor_interval_minutes: int = Field(default=30, ge=5, le=10080)
    collector_type: CollectorType = "mock"
    follower_count: int = Field(default=0, ge=0)
    following_count: int = Field(default=0, ge=0)
    total_like_count: int = Field(default=0, ge=0)
    content_count: int = Field(default=0, ge=0)
    profile_resolved: bool = False

    @field_validator("profile_url", mode="before")
    @classmethod
    def normalize_profile_url_field(cls, value: str) -> str:
        return normalize_profile_url(value)

    @model_validator(mode="after")
    def validate_collector_type(self) -> Self:
        if (
            self.collector_type in {"douyin_public_web", "tikomni_douyin"}
            and self.platform != "douyin"
        ):
            raise ValueError("抖音真实采集器只能用于抖音账号")
        return self


class CreatorProfileResolveRequest(BaseModel):
    platform: Platform
    input_value: str = Field(min_length=1, max_length=1000)


class CreatorProfileResolveResponse(BaseModel):
    platform: Platform
    platform_account_id: str
    platform_display_id: str | None = None
    nickname: str
    profile_url: str
    avatar_url: str | None = None
    bio: str | None = None
    verified_info: str | None = None
    location: str | None = None
    follower_count: int
    following_count: int
    total_like_count: int
    content_count: int
    collector_type: CollectorType
    sec_user_id: str | None = None
    warnings: list[str] = Field(default_factory=list)


class CreatorUpdate(BaseModel):
    nickname: str | None = Field(default=None, min_length=1, max_length=128)
    platform_display_id: str | None = Field(default=None, max_length=128)
    profile_url: str | None = Field(default=None, min_length=1, max_length=500)
    avatar_url: str | None = Field(default=None, max_length=1000)
    bio: str | None = None
    group_name: str | None = Field(default=None, max_length=128)
    tags: list[str] | None = None
    priority: Priority | None = None
    monitor_interval_minutes: int | None = Field(default=None, ge=5, le=10080)
    monitoring_status: MonitoringStatus | None = None
    collector_type: CollectorType | None = None

    @field_validator("profile_url", mode="before")
    @classmethod
    def normalize_profile_url_field(cls, value: str | None) -> str | None:
        return normalize_profile_url(value) if value is not None else None


class CreatorRead(UtcResponseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    platform: str
    platform_account_id: str
    platform_display_id: str | None
    nickname: str
    profile_url: str
    avatar_url: str | None
    bio: str | None
    verified_info: str | None
    location: str | None
    group_name: str | None
    tags: list[str]
    priority: str
    monitor_interval_minutes: int
    monitoring_status: str
    collector_type: str
    collector_version: str | None
    data_quality_status: str
    last_content_status: str
    last_collection_error: str | None
    content_baseline_established_at: datetime | None
    follower_count: int
    following_count: int
    total_like_count: int
    content_count: int
    last_collected_at: datetime | None
    next_collect_at: datetime | None
    consecutive_failures: int
    created_at: datetime
    updated_at: datetime


class CreatorListResponse(BaseModel):
    items: list[CreatorRead]
    total: int
    page: int
    page_size: int


class CreatorSnapshotRead(UtcResponseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    creator_id: int
    follower_count: int
    following_count: int
    total_like_count: int
    content_count: int
    collector_type: str
    data_quality_status: str
    captured_at: datetime


class CollectionRunCreatorRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    nickname: str
    platform: str


class CollectionRunRead(UtcResponseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    creator_id: int
    task_type: str
    status: str
    trigger_source: str
    attempt: int
    collector_type: str | None
    error_type: str | None
    duration_ms: int | None
    started_at: datetime
    finished_at: datetime | None
    error_message: str | None
    result_summary: dict | None
    creator: CollectionRunCreatorRead


class CollectionResult(BaseModel):
    creator: CreatorRead
    snapshot: CreatorSnapshotRead
    run: CollectionRunRead


class CollectionRetryQueued(BaseModel):
    creator_id: int
    task_id: str
    status: Literal["queued"]
    retry_after_seconds: int
    message: str


class CollectionRunListResponse(BaseModel):
    items: list[CollectionRunRead]
    total: int
    page: int
    page_size: int
